File: src/main/java/ms/aurora/loader2/ClientWrapper.scala
package ms.aurora.loader2

import ms.aurora.browser.Browser
import ms.aurora.browser.ContextBuilder.get
import ms.aurora.Messages.getString
import scala.beans.BeanProperty
import java.applet.Applet
import ms.aurora.rt3.Client
import org.apache.log4j.Logger

/**
 * The context of the game client, which acts
 * as some sort of wrapper around the client applet
 * and the loaders / configurations used.
 * @author rvbiljouw
 */
class ClientWrapper {
  private val logger = Logger.getLogger(classOf[ClientWrapper])
  @BeanProperty val browserBaseURL = getString("runescape.url")
  @BeanProperty val browserContext = get.domain(browserBaseURL).build
  @BeanProperty val browser = new Browser(browserContext)
  @BeanProperty val config = new ClientConfig(browser)
  @BeanProperty val loader = new ClientLoader(config)

  def getClient: Client = {
    loader.getApplet.asInstanceOf[Client]
  }

  def getApplet: Applet = {
    loader.getApplet
  }

  def start(): Boolean = {
    config.visit()
    loader.start()
    logger.info("Applet started")
    loader.isLoaded
  }

  def restart() {
    if (stop() && start()) {
      logger.info("Applet restarted")
    } else {
      logger.error("Applet reload failed")
    }
  }

  def stop(): Boolean = {
    loader.stop()
    logger.info("Applet stopped")
    !loader.isLoaded
  }

}

File: src/main/java/ms/aurora/loader/ClientWrapper.java
package ms.aurora.loader;

import ms.aurora.api.methods.web.model.World;
import ms.aurora.browser.Browser;
import ms.aurora.browser.Context;
import ms.aurora.core.Session;
import ms.aurora.rt3.Client;
import org.apache.log4j.Logger;

import java.applet.Applet;

import static ms.aurora.browser.ContextBuilder.get;
import static ms.aurora.gui.Messages.getString;

/**
 * @author rvbiljouw
 */
public class ClientWrapper {
    private static final Logger logger = Logger.getLogger(ClientWrapper.class);
    private final String browserBaseURL = getString("runescape.url");
    private final Context browserContext = get().domain(browserBaseURL).build();
    private final Browser browser = new Browser(browserContext);
    private final ClientConfig config = new ClientConfig(this, browser);
    private final ClientLoader loader = new ClientLoader(config);
    private World world;

    public Client getClient() {
        return (Client) loader.getApplet();
    }

    public Applet getApplet() {
        return loader.getApplet();
    }

    public boolean start() {
        config.visit();
        loader.start();
        return loader.isLoaded();
    }

    public boolean stop() {
        loader.stop();
        return !loader.isLoaded();
    }

    public void restart() {
        if (stop() && start()) {
            logger.info("Applet restarted");
        } else {
            logger.error("Applet reload failed.");
        }
    }

    public World getWorld() {
        return world;
    }

    public boolean setWorld(Session session, World world) {
        this.world = world;

        if (loader.isLoaded()) {
            session.getScriptManager().pause();
            restart();
            session.update();
            session.getScriptManager().resume();
            return true;
        }
        return false;
    }
}
